src/main/java/rx/internal/operators/OperatorEagerConcatMap.java

```diff
@@ -0,0 +1,299 @@
+        final int bufferSize;
+        final Subscriber<? super R> actual;
+        
+        final LinkedList<EagerInnerSubscriber<R>> subscribers;
+        
+        volatile boolean done;
+        Throwable error;
+        
+        volatile boolean cancelled;
+        
+        final AtomicInteger wip;
+        private EagerOuterProducer sharedProducer;
+        
+        public EagerOuterSubscriber(Func1<? super T, ? extends Observable<? extends R>> mapper, int bufferSize,
+                Subscriber<? super R> actual) {
+            this.mapper = mapper;
+            this.bufferSize = bufferSize;
+            this.actual = actual;
+            this.subscribers = new LinkedList<EagerInnerSubscriber<R>>();
+            this.wip = new AtomicInteger();
+        }
+        
+        void init() {
+            sharedProducer = new EagerOuterProducer(this);
+            add(Subscriptions.create(new Action0() {
+                @Override
+                public void call() {
+                    cancelled = true;
+                    if (wip.getAndIncrement() == 0) {
+                        cleanup();
+                    }
+                }
+            }));
+            actual.add(this);
+            actual.setProducer(sharedProducer);
+        }
+        
+        void cleanup() {
+            List<Subscription> list;
+            synchronized (subscribers) {
+                list = new ArrayList<Subscription>(subscribers);
+                subscribers.clear();
+            }
+            
+            for (Subscription s : list) {
+                s.unsubscribe();
+            }
+        }
+        
+        @Override
+        public void onNext(T t) {
+            Observable<? extends R> observable;
+            
+            try {
+                observable = mapper.call(t);
+            } catch (Throwable e) {
+                Exceptions.throwOrReport(e, actual, t);
+                return;
+            }
+            
+            EagerInnerSubscriber<R> inner = new EagerInnerSubscriber<R>(this, bufferSize);
+            if (cancelled) {
+                return;
+            }
+            synchronized (subscribers) {
+                if (cancelled) {
+                    return;
+                }
+                subscribers.add(inner);
+            }
+            if (cancelled) {
+                return;
+            }
+            observable.unsafeSubscribe(inner);
+            drain();
+        }
+        
+        @Override
+        public void onError(Throwable e) {
+            error = e;
+            done = true;
+            drain();
+        }
+        
+        @Override
+        public void onCompleted() {
+            done = true;
+            drain();
+        }
+        
+        void drain() {
+            if (wip.getAndIncrement() != 0) {
+                return;
+            }
+            int missed = 1;
+            
+            final AtomicLong p = sharedProducer;
```

Comment:
p is not being used in the sense of a producer but rather in its sense of the request count. I'd like to see this renamed to `requested` or something similar in keeping with naming in other operators.